pkg/network/ebpf/c/protocol-classification-helpers.h

```diff
@@ -0,0 +1,269 @@
+}
+
+// Given protocols for the socket connection tuple, and the inverse skb connection tuple, the function returns
+// the final protocol among the two.
+// If the sock_tup_protocol is unclassified, then it does not matter what's the value of the inverse_skb_tup_protocol,
+// we will take it. If the inverse_skb_tup_protocol is unclassified as well, then it does not matter which "unclassified"
+// we choose. If it is unknown or classified, then we should choose it.
+// If the sock_tup_protocol is unknown, then we take the inverse_skb_tup_protocol if it is classified or unknown.
+// If both are unknown, then it does not matter which "unknown" we choose. If the inverse_skb_tup_protocol is classified,
+// then for sure we should choose it.
+// On any other case take sock_tup_protocol.
+static __always_inline protocol_t choose_protocol(protocol_t sock_tup_protocol, protocol_t inverse_skb_tup_protocol) {
+    if ((sock_tup_protocol == PROTOCOL_UNCLASSIFIED) ||
+        (sock_tup_protocol == PROTOCOL_UNKNOWN && inverse_skb_tup_protocol != PROTOCOL_UNCLASSIFIED)) {
+        return inverse_skb_tup_protocol;
+    }
+
+    // On any other case, we give the priority to the classified protocol for the socket connection tuple
+    return sock_tup_protocol;
+}
+
+// Replaces the source and dest fields (addresses and ports).
+static __always_inline void invert_conn_tuple(conn_tuple_t *original_conn, conn_tuple_t *output_conn) {
+    output_conn->saddr_h = original_conn->daddr_h;
+    output_conn->saddr_l = original_conn->daddr_l;
+    output_conn->daddr_h = original_conn->saddr_h;
+    output_conn->daddr_l = original_conn->saddr_l;
+    output_conn->sport = original_conn->dport;
+    output_conn->dport = original_conn->sport;
+    output_conn->metadata = original_conn->metadata;
+    output_conn->pid = original_conn->pid;
+    output_conn->netns = original_conn->netns;
+}
+
+// A shared implementation for the runtime & prebuilt socket filter that classifies the protocols of the connections.
+static __always_inline void protocol_classifier_entrypoint(struct __sk_buff *skb) {
+    skb_info_t skb_info = {0};
+    conn_tuple_t skb_tup = {0};
+
+    // Exporting the conn tuple from the skb, alongside couple of relevant fields from the skb.
+    if (!read_conn_tuple_skb(skb, &skb_info, &skb_tup)) {
+        return;
+    }
+
+    // We process a non empty TCP packets, rather than that - we skip the packet.
+    if (!should_process_packet(skb, &skb_info, &skb_tup)) {
+        return;
+    }
+
+    // Making sure we've not processed the same tcp segment, which can happen when a single packet travels different
+    // interfaces.
+    if (has_sequence_seen_before(&skb_tup, &skb_info)) {
+        return;
+    }
+
+    conn_tuple_t *cached_sock_conn_tup_ptr = bpf_map_lookup_elem(&skb_conn_tuple_to_socket_conn_tuple, &skb_tup);
+    if (cached_sock_conn_tup_ptr == NULL) {
+        return;
+    }
+
+    conn_tuple_t cached_sock_conn_tup = *cached_sock_conn_tup_ptr;
```

Comment:
What do you think about making a new type for this, like `minimized_conn_tuple_t`. And embedding this inside `conn_tuple_t`. Then we won't have to worry about setting netns and pid to 0 everywhere.

Reply:
good suggestion, I'll make the changes